lib/pbench/server/api/resources/query_apis/iteration_samples.py

```diff
@@ -0,0 +1,271 @@
+            ),
+        )
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this run id, and determine whether the
+        authenticated user has READ access to this dataset. (Currently, this
+        means the authenticated user is the owner of the dataset, or has ADMIN
+        role.)
+        If the user has authorization to read the dataset, return the Dataset
+        object as CONTEXT so that the postprocess operation can proceed to get all
+        the iteration aggregations.
+        """
+        run_id = client_json.get("run_id")
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.error("Dataset with Run ID {} not found", run_id)
+            abort(HTTPStatus.NOT_FOUND, message="Dataset not found")
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                "Dataset owner ID {} cannot be found in Users", dataset.owner_id
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For Iteration samples, we check authorization against the ownership of the
+        # dataset that was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, dataset.access)
+
+        # The dataset exists, and authenticated user has enough access so continue
+        # the operation with the appropriate CONTEXT.
+        return {"dataset": dataset}
+
+
+class IterationSampleRows(IterationSamples):
+    """
+    Note: This is a first part of queryIterationSamples API that returns only unique rows
+    for each subdocument.
+
+    This class inherits the common IterationSamples class and builds an aggregated query
+    against a user supplied run id to compile the filter header which shows every unique
+    field in the 'benchmark', 'sample', 'iterations' and other subdocuments.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config, logger,
+        )
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        """
+        Construct a pbench search query for aggregating unique values for keyword fields in the
+        'benchmark', 'sample', 'iterations' and other sub-documents of result-data-sample index document
+        that belong to the given run id within a specified date range.
+        {
+            "start": "start-time",
+            "end": "end-time",
+            "run_id": "run id string"
+        }
+        json_data: JSON dictionary of type-normalized parameters
+            "start" and "end" are string representation of datetime objects to search within.
+            "run_id": String representation of run id
+        """
+        start = json_data.get("start")
```

Comment:
It can be in the metadata table since it has a foreign key that points to the dataset entry. So if we can get the dataset from run_id, we can find the metadata related to that dataset and get the index stored at `server.index-map`